includes/content.php

```diff
@@ -179,96 +191,139 @@ function pmpro_search_filter($query)
-        global $pmpro_my_cats;
-		$pmpro_my_cats = array();
-
-        if( $levels ) {
-            foreach( $levels as $key => $level ) {
-                $member_cats = pmpro_getMembershipCategories($level->id);
-                $pmpro_my_cats = array_unique(array_merge($pmpro_my_cats, $member_cats));
-            }
-        }
-		
-        //get hidden cats
-        if( ! empty( $pmpro_my_cats ) ) {
-			$sql = "SELECT category_id FROM $wpdb->pmpro_memberships_categories WHERE category_id NOT IN(" . implode(',', array_map( 'esc_sql', $pmpro_my_cats ) ) . ")";
-		} else {
-			$sql = "SELECT category_id FROM $wpdb->pmpro_memberships_categories";
-		}							
-        $hidden_cat_ids = array_values(array_unique($wpdb->get_col($sql)));
-		
-        //make this work
-        if( $hidden_cat_ids ) {
-
-			// Get all registered category ID's so we may remove the ones we need to hide. This is to support posts that may belong to multiple categories and only one of them is hidden.
-			$all_cat_ids = get_terms( array(
-				'taxonomy' => 'category',
-				'fields'   => 'ids',
-				'get'      => 'all',
-			));
-
-			// Show only these categories to members.
-			$query->set( 'category__in', array_merge( $query->get( 'category__in' ), array_diff( $all_cat_ids, $hidden_cat_ids ) ) );
-						
-			//filter so posts in this member's categories are allowed
-			add_action( 'posts_where', 'pmpro_posts_where_unhide_cats' );
-		}
-    }
+	/**
+	 * Okay. We're going to filter the search results.
+	 * Some explaination of what we're doing.
+	 * A = All posts hidden by level.
+	 * B = All posts the current user has access to by level.
+	 * C = All posts hidden by category.
+	 * D = All posts the current user has access to by category.
+	 * 
+	 * Then hidden posts = (A - B) + (C - D)
+	 * 
+	 * We calculate A-B and call it $hidden_post_ids.
+	 * We calculate C-D and call it $cat_hidden_post_ids.
+	 * 
+	 * Then we merge and add to post__not_in.
+	 */
+	
+	/**
+	 * Get hidden post IDs.
+	 * Note this also gets IDs for posts and other custom post types
+	 * that might be protected through the PMPro CPT Add On or
+	 * other plugins or custom code.
+	 */
+	// First, figure out if their is a current user with levels.
+	if( ! empty( $current_user->ID ) ) {
+		$levels = pmpro_getMembershipLevelsForUser($current_user->ID);
+	} else {
+		$levels = false;
+	}
+	if( ! empty( $levels ) ) {
+		$level_ids = wp_list_pluck( $levels, 'ID' );
+	}
+
+	// Now get posts hidden by level.
+	if( ! empty( $level_ids ) ) {
+		// All hidden pages, minus the ones I have access to.
+		$sql = "SELECT mp.page_id
+				FROM {$wpdb->pmpro_memberships_pages} mp
+					LEFT JOIN {$wpdb->posts} p ON mp.page_id = p.ID
+				WHERE mp.page_id NOT IN(
+					SELECT page_id
+					FROM {$wpdb->pmpro_memberships_pages}
+					WHERE membership_id IN(" . implode(',', array_map( 'esc_sql', $level_ids ) ) . ")
+				) AND p.post_type IN('" . implode( "', '", array_map( 'esc_sql', $pmpro_search_filter_post_types ) ) . "')";
```

Comment:
Would `SELECT DISTINCT` be more performant than just using `SELECT` and then calling `array_unique`, or is the difference minimal?

Reply:
Good idea. In general, it's better to do as much as possible in the query.